hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -1231,34 +1241,46 @@ int SaraNcpClient::initReady(ModemState state) {
 }
 
-int SaraNcpClient::checkRuntimeState(ModemState& state) {
-    // Assume we are running at the runtime baudrate
-    unsigned runtimeBaudrate = ncpId() == PLATFORM_NCP_SARA_R410 ? UBLOX_NCP_RUNTIME_SERIAL_BAUDRATE_R4 :
-            UBLOX_NCP_RUNTIME_SERIAL_BAUDRATE_U2;
+bool SaraNcpClient::checkRuntimeStateMuxer(unsigned int baudrate) {
+    // Assume we are running at the given baudrate
+    CHECK(serial_->setBaudRate(baudrate));
 
-    CHECK(serial_->setBaudRate(runtimeBaudrate));
+    // Attempt to check whether we are already in muxer state
+    // This call will automatically attempt to open/reopen channel 0 (main)
+    // It should timeout within T2 * N2, which is 300ms * 3 ~= 1s
+    bool stop = true;
+    SCOPE_GUARD({
+        if (stop) {
+            muxer_.stop();
+        }
+    });
+    if (!initMuxer()) {
+        muxer_.setAckTimeout(gsm0710::proto::DEFAULT_T1);
+        muxer_.setControlResponseTimeout(gsm0710::proto::DEFAULT_T2);
+        LOG(TRACE, "Initialized muxer @ %u baudrate", baudrate);
+        if (!muxer_.start(true /* initiator */, false /* don't open channel 0 */)) {
+            if (!muxer_.forceOpenChannel(0) && !muxer_.ping()) {
+                LOG(TRACE, "Resumed muxed session");
+                stop = false;
+                return true;
+            }
+        }
+    }
+    return false;
+}
 
+int SaraNcpClient::checkRuntimeState(ModemState& state, unsigned runtimeBaudrate) {
     // Feeling optimistic, try to see if the muxer is already available
     if (!muxer_.isRunning()) {
         LOG(TRACE, "Muxer is not currently running");
-        // Attempt to check whether we are already in muxer state
-        // This call will automatically attempt to open/reopen channel 0 (main)
-        // It should timeout within T2 * N2, which is 300ms * 3 ~= 1s
-        bool stop = true;
-        SCOPE_GUARD({
-            if (stop) {
-                muxer_.stop();
-            }
-        });
-        if (!initMuxer()) {
-            muxer_.setAckTimeout(gsm0710::proto::DEFAULT_T1);
-            muxer_.setControlResponseTimeout(gsm0710::proto::DEFAULT_T2);
-            LOG(TRACE, "Initialized muxer");
-            if (!muxer_.start(true /* initiator */, false /* don't open channel 0 */)) {
-                if (!muxer_.forceOpenChannel(0) && !muxer_.ping()) {
-                    LOG(TRACE, "Resumed muxed session");
-                    stop = false;
-                }
+        if (ncpId() != PLATFORM_NCP_SARA_R410) {
+            checkRuntimeStateMuxer(runtimeBaudrate);
+        } else {
+            LOG_DEBUG(TRACE, "Attempt to start/resume muxer at %u baud", UBLOX_NCP_RUNTIME_SERIAL_BAUDRATE_R4);
+            int ret = checkRuntimeStateMuxer(UBLOX_NCP_RUNTIME_SERIAL_BAUDRATE_R4);
+            if (!ret) {
+                LOG_DEBUG(TRACE, "Attempt to start/resume muxer at %u baud", UBLOX_NCP_DEFAULT_SERIAL_BAUDRATE);
+                checkRuntimeStateMuxer(UBLOX_NCP_DEFAULT_SERIAL_BAUDRATE);
```

Comment:
Runtime baudrate should be restored if this fails, otherwise https://github.com/particle-iot/device-os/pull/2146/files#diff-ece868c5f58dddb291c6b739de0da632R1333 will check at default baudrate and then further down we'll check again at default baudrate.